google/cloud/sql/connector/refresh_utils.py

```diff
@@ -229,3 +231,34 @@ async def _is_valid(task: asyncio.Task) -> bool:
         # supress any errors from task
         logger.debug("Current instance metadata is invalid.")
     return False
+
+
+def _downscope_credentials(
+    credentials: Credentials,
+    scopes: List[str] = ["https://www.googleapis.com/auth/sqlservice.login"],
+) -> Credentials:
+    """Generate a down-scoped credential.
+
+    :type credentials: google.auth.credentials.Credentials
+    :param credentials
+        Credentials object used to generate down-scoped credentials.
+
+    :type scopes: List[str]
+    :param scopes
+        List of Google scopes to include in down-scoped credentials object.
+
+    :rtype: google.auth.credentials.Credentials
+    :returns: Down-scoped credentials object.
+    """
+    if isinstance(credentials, Scoped):
+        scoped_creds = credentials.with_scopes(scopes=scopes)
+    # authenticated user credentials can not be re-scoped
+    else:
+        # create shallow copy to not overwrite scopes on default credentials
+        scoped_creds = copy.copy(credentials)
+        scoped_creds._scopes = scopes
+    # downscoped credentials require refresh
+    if not scoped_creds.valid:
+        request = google.auth.transport.requests.Request()
+        scoped_creds.refresh(request)
```

Comment:
We should refresh every time before we create the ephemeral cert, so that it's duration is as long as possible. I think it might also be good to move this logic out of the function and into the ephemeral cert func for this reason.

Reply:
We already do refresh before ephemeral cert API call ([code snippet](https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/blob/main/google/cloud/sql/connector/refresh_utils.py#L169-L171)). The reason for this refresh is that once you re-scope a credential and create a new credential object, it seems to become invalid and thus requires a new refresh.

Reply:
we shouldn't need to refresh here if it refreshes there then?

Reply:
I think we need to refresh both credentials don't we? We need to still refresh the default credential so its token is valid in the header, we also need to refresh the down-scoped credential for it's token to be valid. Thoughts?

Reply:
I understand, removed the `or enable_iam_auth` from the if statement